### ISSUE TITLE
optimizer: redundant join not removed

### ISSUE THREAD

**Issue by asenac** [MaterializeInc/materialize#6988]
```
materialize=> explain with a(a) as (select b from a group by b) select * from a, a b where a.a = b.a;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.a (u707)+
 | Filter !(isnull(#1))           +
 | Distinct group=(#1)            +
 | ArrangeBy (#0)                 +
                                  +
 %1 =                             +
 | Get materialize.public.a (u707)+
 | Filter !(isnull(#1))           +
 | Distinct group=(#1)            +
 | ArrangeBy (#0)                 +
                                  +
 %2 =                             +
 | Join %0 %1 (= #0 #1)           +
 | | implementation = DeltaQuery  +
 | |   delta %0 %1.(#0)           +
 | |   delta %1 %0.(#0)           +
 | | demand = (#0)                +
 | Project (#0, #0)               +
 
(1 row)

materialize=> 
```

The join above could be removed since both operands come from the same source and are joined through their unique key.

**Comment by uce** [comment]
This will probably be handled by QGM (#7457). @asenac Can you confirm?

**Comment by asenac** [comment]
This can be handled either in the QGM or in Mir if we had a Common Sub-Expression detection transform (#7046).

Simple cases like this one could be also be a side-effect of lowering CTEs as CSEs as the QGM predicate does already:

```
plan for with a(a) as (select column1 from b) select b.* from a b, a c:
%0 = Let l0 =
| Constant ()

%1 =
| Get ? (u0)

%2 = Let l1 =
| Join %0 %1
| | implementation = Unimplemented
| Project (#0..#2)
| Filter 
| Map #0
| Project (#3)
| Project (#0)
| Filter 
| Map #0
| Project (#1)

%3 =
| Get %2 (l1)

%4 =
| Join %3 %2
| | implementation = Unimplemented
| Project (#0, #0)
| Filter 
| Map #0
| Project (#2)
```

**Comment by asenac** [comment]
#7715 fixed this issue:

```
materialize=> explain with a(a) as (select b from a group by b) select * from a, a b where a.a = b.a;
         Optimized Plan          
---------------------------------
 %0 =                           +
 | Get materialize.public.a (u9)+
 | Filter !(isnull(#1))         +
 | Project (#1)                 +
 | Distinct group=(#0)          +
 | Project (#0, #0)             +
 
(1 row)

materialize=> 
```

Not sure if an explicit test was added for this case. Will check.

**Comment by frankmcsherry** [comment]
This is reported as fixed in #7715 so closing.
